stan/math/prim/prob/bernoulli_logit_lpmf.hpp

```diff
@@ -43,38 +49,44 @@ return_type_t<T_prob> bernoulli_logit_lpmf(const T_n& n, const T_prob& theta) {
   }
 
   T_partials_return logp(0.0);
-  operands_and_partials<T_prob> ops_partials(theta);
+  operands_and_partials<T_theta_ref> ops_partials(theta_ref);
 
-  scalar_seq_view<T_n> n_vec(n);
-  scalar_seq_view<T_prob> theta_vec(theta);
-  size_t N = max_size(n, theta);
+  const auto& theta_val = value_of(theta_ref);
+  const auto& theta_arr = as_array_or_scalar(theta_val);
+  const auto& n_double = value_of_rec(n_ref);
 
-  for (size_t n = 0; n < N; n++) {
-    const T_partials_return theta_dbl = value_of(theta_vec[n]);
-
-    const int sign = 2 * n_vec[n] - 1;
-    const T_partials_return ntheta = sign * theta_dbl;
-    const T_partials_return exp_m_ntheta = exp(-ntheta);
+  auto signs = to_ref_if<!is_constant<T_prob>::value>(
+      (2 * as_array_or_scalar(n_double) - 1));
+  T_partials_array ntheta;
+  if (is_vector<T_n>::value || is_vector<T_prob>::value) {
```

Comment:
This is an expensive way to assign signs for autodiff because it introduces new `sign * theta_arr` expressions.   The arithmetic is integer, so that's cheap if not completely parallelizable in the CPU.

The alternative is a conditional, but that's also expensive in terms of the actual branching.  A better thing to do with complete control would be to organize the data into two groups that we could just handle directly, but we can't assume that with this function.

So I'm OK leaving this as is because I don't know if the alternative is any more efficient.

Reply:
Up to you.  This one was optional.  I'm going to mark approved and I can re-review if it changes or you can merge.

Reply:
It turns out multiplication is a tiny bit faster.